src/merger_tree.cpp

```diff
@@ -0,0 +1,175 @@
+                     count_found, trackids_found, world.Communicator);
+  }
+
+  // Now identify a descendant for each lost subhalo in the DescendantTracerIds map
+  std::vector<std::pair<HBTInt,HBTInt>> DescendantTrackIds;
+  {
+    HBTInt count_found_offset = 0;
+    HBTInt trackids_found_offset = 0;
+    DescendantTrackIds.reserve(DescendantTracerIds.size());
+    for (const auto &p : DescendantTracerIds) {
+
+      // TrackId of the lost subhalo
+      const HBTInt &LostTrackId = p.first;
+    
+      // Vector of tracer particle IDs from the lost subhalo
+      const std::vector<HBTInt> &LostTracerIds = p.second;
+    
+      // Make a map where the keys are TrackIds of the possible descendants
+      // and the values are counts of how many particles went to each one.
+      std::map<HBTInt,HBTInt> nr_desc_particles;
+      // Loop over tracers to find for this lost subhalo
+      for(HBTInt i=0; i<LostTracerIds.size(); i+=1) {
+        // Loop over number of times each tracer was found
+        assert(count_found_offset < count_found.size());
+        for(HBTInt j=0; j<count_found[count_found_offset]; j+=1) {
+          assert(trackids_found_offset < trackids_found.size());
+          const HBTInt desc_trackid = trackids_found[trackids_found_offset];
+          if(nr_desc_particles.count(desc_trackid) == 0) {
+            // This is the first particle going to this descendant
+            nr_desc_particles[desc_trackid] = 1;
+          } else {
+            // Increment the number of particles going to this descendant
+            nr_desc_particles[desc_trackid] += 1;          
+          }
+          trackids_found_offset += 1;
+        }
+        count_found_offset += 1;
+      }
+      // Now check which subhalo received the largest number of particles.
+      // This is the TrackId key with the largest associated value in nr_desc_particles.
+      HBTInt max_trackid = SpecialConst::NullTrackId;
+      HBTInt max_count = 0;
+      for(const auto &p: nr_desc_particles) {
+        const HBTInt desc_trackid = p.first;
+        const HBTInt desc_count = p.second;
+        if(desc_count > max_count) {
+          max_trackid = desc_trackid;
+          max_count = desc_count;
+        }
+      }
+      // Store the result in the same order as the DescendantTracerIds map keys
+      std::pair<HBTInt,HBTInt> desc_pair(LostTrackId, max_trackid);
+      DescendantTrackIds.push_back(desc_pair);
+    }
+    assert(count_found_offset==count_found.size());
+    assert(trackids_found_offset==trackids_found.size());
+  }
+
+  // Now we need to update DescendantTrackId in the Subhalo_t structs.
+  // Here we assume that subhalos have not migrated between MPI ranks since
+  // Unbind() was called, so all entries in the DescendantTracerIds map
+  // correspond to subhalos which are stored on this MPI rank.
+  {
+    // Make a sorting index for the subhalos
+    std::vector<HBTInt> subhalo_trackid;
```

Comment:
Might be better to directly initialise/reserve the subhalo_trackid vector with the number of elements we need (we know the number of subhaloes) since we are currently re-sizing the vector as we add TrackId values.

Reply:
Done.